sdk/python/lib/pulumi/provider/server.py

```diff
@@ -0,0 +1,189 @@
+"""Define gRPC plumbing to expose a custom user-defined `Provider`
+instance as a gRPC server so that it can be used as a Pulumi plugin.
+
+"""
+
+from typing import Dict, List, Optional
+import argparse
+import asyncio
+import sys
+
+import grpc
+import grpc.aio
+import google.protobuf.struct_pb2 as struct_pb2
+
+from pulumi.provider.provider import Provider, ConstructResult
+from pulumi.runtime import proto, rpc
+from pulumi.runtime.proto import provider_pb2_grpc, ResourceProviderServicer
+from pulumi.runtime.proto.provider_pb2 import ConstructRequest
+import pulumi
+import pulumi.resource
+import pulumi.runtime.config
+import pulumi.runtime.settings
+
+
+# _MAX_RPC_MESSAGE_SIZE raises the gRPC Max Message size from `4194304` (4mb) to `419430400` (400mb)
+_MAX_RPC_MESSAGE_SIZE = 1024 * 1024 * 400
+_GRPC_CHANNEL_OPTIONS = [('grpc.max_receive_message_length', _MAX_RPC_MESSAGE_SIZE)]
+_GRPC_WORKERS = 4
+
+
+class ProviderServicer(ResourceProviderServicer):
+    """Implements a subset of `ResourceProvider` methods to support
+    `Construct` and other methods invoked by the engine when the user
+    program creates a remote `ComponentResource` (with `remote=true`
+    in the constructor).
+
+    See `ResourceProvider` defined in `provider.proto`.
+
+    """
+
+    engine_address: str
+    provider: Provider
+    args: List[str]
+
+    async def Construct(self, request: ConstructRequest, context):  # pylint: disable=invalid-overridden-method
+        assert isinstance(request, ConstructRequest), \
+            f'request is not ConstructRequest but is {type(request)} instead'
+
+        # Should we take a lock so Construct calls are never executing
+        # in parallel? Node SDK does this. Also, should we set these
+        # settings back after we are done?
+
+        pulumi.runtime.settings.reset_options(
+            project=_empty_as_none(request.project),
+            stack=_empty_as_none(request.stack),
+            parallel=_empty_as_none(request.parallel),
+            engine_address=self.engine_address,
+            monitor_address=_empty_as_none(request.monitorEndpoint),
+            preview=request.dryRun)
+
+        pulumi.runtime.config.set_all_config(dict(request.config))
+
+        inputs = self._construct_inputs(request)
+
+        result = self.provider.construct(name=request.name,
+                                         resource_type=request.type,
+                                         inputs=inputs,
+                                         options=self._construct_options(request))
+
+
+        response = await self._construct_response(result)
+
+        ## should we do runtime.disconnect()? where is this
```

Comment:
Unresolved Q2

Reply:
The main thing we're using `runtime.disconnect()` for in Node.js is to wait for in-flight RPC calls to finish before returning from `Construct`, to ensure any pending `registerResource` calls (for a component's child resources) have completed before returning.

We do the same for `Go` by calling `waitForRPCs`: https://github.com/pulumi/pulumi/blob/34eae7e7e3ddae475c52f2e248b1eabb23d62cf7/sdk/go/pulumi/provider.go#L687-L690

I'm less familiar with this part of the codebase, but I think we may need to run the user's `construct` within `run_pulumi_func`, which looks like it'll handle this:

https://github.com/pulumi/pulumi/blob/ee2f65510bd8129adffb18d7e198bf12c1b84fee/sdk/python/lib/pulumi/runtime/stack.py#L41

Reply:
Ack will make it happen in Python.

Curious, what kind of user code would lead to a situation where there are outstanding RPCs after provider.construct has returned, and especially after we're waited on the constructed component's URN Output. Something along the lines of the code creating child components and not awaiting their URNs?

Reply:
> Something along the lines of the code creating child components and not awaiting their URNs?

Yeah

Reply:
Not good, adding `await wait_for_rpcs()` makes the example test not terminate. I'm trying to get detailed logs out to understand what is going on.

Reply:
OK so, as it turns out, wait_for_rpcs() waits for     _all_tasks = asyncio.Task.all_tasks

```
all_tasks(loop=None) method of builtins.type instance
    Return a set of all tasks for an event loop.
    
    By default all tasks for the current event loop are returned.
```

That I think is a bug - this time it picked up AIO tasks related to running the gRPC server and started blocking on those.

I think it is a bug to ever poke into all_tasks as we want user code to be able to schedule their own tasks without interference on the shared event loop.

Do we have a dedicated event loop to segregate?

 However it might not be possible to refactor swiftly. Please advice what's a good way forward. @komalali

Reply:
...where/what is `wait_for_rpcs()`?

In any case, these are the dragons I was hinting at [up here](https://github.com/pulumi/pulumi/pull/6715#discussion_r611876606). `run_pulumi_func` has some very odd behavior that I haven't been able to tease apart which is why I've largely left the whole [RPC_MANAGER stuff](https://github.com/pulumi/pulumi/blob/master/sdk/python/lib/pulumi/runtime/stack.py#L48-L58) alone because I also ran into infinite hangs when trying to address it. 

> Do we have a dedicated event loop to segregate?
Unfortunately, we tend to just [grab the existing event loop](https://github.com/pulumi/pulumi/blob/master/sdk/python/lib/pulumi/runtime/sync_await.py#L90-L98) so we don't have a dedicated one.

You have found yourself in some of the hairiest python code in our codebase. Maybe we can do a quick call to talk through options?

Reply:
Yeah that sounds good. I am tempted to burn it with FIRE so talking is good.

Reply:
https://github.com/pulumi/pulumi/issues/6762 we logged this, I will proceed to try to isolate only the code that waits on RPCs, that might work for the purposes of this PR. Revisit the 6761 later.

Reply:
Waiting just for RPCs does not lock, great news.